### DISCUSSION TITLE
Ability to revalidate the literal route segment

### DISCUSSION THREAD

**Discussion by nrsimha** [vercel/next.js#53154] [Ideas]
### Goals

To be able to revalidate the literal route segment (e.g. post/123456) without need of revalidating all products

### Background

Let say site has 1M of post. If one post is edited and all other are unchanged it will be nice to be able to revalidate just one which has been edited and keep remaining 999,999 without change and without need of rebuild them again.

Currently revalidatePath doesn't support this, so revalidatePath('/post/[id]') will revalidate all 1M of posts.

### Proposal

Add support to revalidatePath to support literal route segment (e.g. post/123456) to be revalidated. So when revalidatePath('/post/123456') is run it will revalidate only post with id of 123456.

**Comment by icyJoseph** [comment]
Is it possible for your application to use `revalidateTag`? I think the purpose of revalidateTag is to give a bit more of granular control over which piece of data is revalidated, precisely for a use case as you describe. Your tag would be something like `post/123`, and you pass it to fetch. 

```ts
fetch(url, { next: { tags: [...] } });
```

**Reply by nrsimha** [reply]
I added following code in app/revalidate/route.js, but it doesn't revalidate:
```
  fetch("http://localhost:3000/post/123456/", {
    next: { tags: ["post-123456"] },
  });

  revalidateTag("post-123456");
```

Doesn't that work only for external pages with fetch or does it suppose to revalidate next.js SSG pages?

**Reply by icyJoseph** [reply]
Mmmm if that did not work, then that's it I think, I'd have expected it to refresh the cache associated with those tags, and revalidate pages that depend on them... Did you refresh the page? Like, without server actions I don't know if there's a way to revalidate without reloading the page.

**Reply by nrsimha** [reply]
Of course that I refreshed page.

Previously if you used /pages directory it was possible to revalidate specific page, but in /app directory it is not possible.

In docs for arguments for `revalidatePath` function (https://nextjs.org/docs/app/api-reference/functions/revalidatePath#parameters) it is written: 

```
revalidatePath(path: string): void;

path: A string representing the filesystem path associated with the data you want to revalidate. This is not the literal route segment (e.g. /product/123) but instead the path on the filesystem (e.g. /product/[id]).
```

Which is really not practical if you have 1,000,000 pages and only one of them needs to be revalidated.

**Reply by icyJoseph** [reply]
Yeah I know that, but I had expected the `revalidateTags` to cover for the use case. 

Though the pages revalidate was called on the API route response object. So it is most likely an entirely different implementation. 

For completeness sake, since you've got a setup already. What happens if you create an API route under pages, and do the res.revalidate (of an app route page) as it was done in that stack?

**Reply by nrsimha** [reply]
Good idea, but res.revalidate unfortunately didn't revalidate app route pages.

**Reply by nrsimha** [reply]
I hope there will be some solution as if you have project where million of pages uses [...path] or [...slug] it is really bad to not be able to revalidate just one of them.

**Reply by icyJoseph** [reply]
Oh darn... Well, it might be time to open an issue. Perhaps the revalidate path function should take a second argument or something to make it be exact or loose, idk

**Reply by justkeen** [reply]
Yes please, I require this as well. No point to revalidate all when just one segment changes.

**Comment by salex97** [comment]
According to the documentation the problem is now solved.

The revalidatePath function now lets us revalidate the literal route segment from a dynamic path without revalidating all other dynamic segments aswell.

https://nextjs.org/docs/app/api-reference/functions/revalidatePath

**Reply by lucasbajoua** [reply]
The phrasing there is confusing, it sounds like all included routes in the dynamic segment still will be invalidated but only on the next request for them.
